### PR TITLE
Add test for the integration with Solidus backend

### LINKED ISSUE #150: Add test for the integration with `solidus_backend`
## What 

QA the backend integration, make sure all the necessary partials and behaviors are there and work as expected.
Add missing specs to cover any regular or edge case.

```[tasklist]
### Test cases
- [x] refunding a payment
- [x] voiding a payment
- [x] canceling an order
- [ ] https://github.com/solidusio/solidus_stripe/issues/266
- [x] navigating payments
- [x] inspecting/rendering log entries
```

## Why

It's likely that once finished we'll still have some edge cases to cover, this is meant as a final QA that should be done before releasing.

### PR DESCRIPTION
**Pull request by rainerdema**
## Summary

Closes #150 

The tests cover the following cases:
| Description | Context | Expected Result | Pending |
|--|:--:|:--|:--:|
| Navigates to the payments page |  with successful payment operations |  Payment is displayed | no |
| Displays log entries for a payment |  with successful payment operations |  Log entries are displayed | no |
| Captures an authorized payment successfully |  with successful payment operations |  Payment is successfully captured | no |
| Voids a payment from an order |  with successful payment operations |  Payment is successfully voided | no |
| Refunds a payment from an order |  with successful payment operations |  Payment is successfully refunded | no |
| Partially refunds a payment from an order |  with successful payment operations |  Payment is partially refunded | no |
| Cancels an order with captured payment |  with successful payment operations |  Payment is successfully voided | yes |
| Cancels an order with authorized payment |  with successful payment operations |  Payment is successfully voided | yes |
| Fails to capture a payment due to incomplete 3D Secure authentication |  with failed payment operations |  Payment capture fails and notification is displayed | no |

## Checklist

Check out our [PR guidelines](https://github.com/solidusio/.github/blob/master/CONTRIBUTING.md#pull-request-guidelines) for more details.

The following are mandatory for all PRs:

- [x] I have written a thorough PR description.
- [x] I have kept my commits small and atomic.
- [x] [I have used clear, explanatory commit messages](https://github.com/solidusio/.github/blob/main/CONTRIBUTING.md#writing-good-commit-messages).

The following are not always needed:

- 📖 I have updated the README to account for my changes.
- 📑 I have documented new code [with YARD](https://www.rubydoc.info/gems/yard/file/docs/Tags.md).
- 🛣️ I have opened a PR to update the [guides](https://github.com/solidusio/edgeguides).
- ✅ I have added automated tests to cover my changes.
- 📸 I have attached screenshots to demo visual changes.

### CODE REVIEW

File: spec/support/solidus_stripe/backend_test_helper.rb [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+          number: card_number,
+          exp_month: 12,
+          exp_year: (Time.zone.now.year + 1),
+          cvc: '123',
+        },
+      })
+    end
+  end
+
+  def create_payment_intent(stripe_payment_method_id)
```
Comment (waiting-for-dev) [inline spec/support/solidus_stripe/backend_test_helper.rb]:
```suggestion
  def create_stripe_payment_intent(stripe_payment_method_id)
```

For consistency :slightly_smiling_face:

**Review by waiting-for-dev — APPROVED** [review]
Cool :raised_hands:

File: app/models/solidus_stripe/gateway.rb [hunk 2]
```diff
@@ -98,7 +98,7 @@ def purchase(amount_in_cents, source, options = {})
     end
 
     # Voids a previously authorized transaction, releasing the funds that are on hold.
-    def void(payment_intent_id, _source, _options = {})
```
Comment (elia) [inline app/models/solidus_stripe/gateway.rb:113]: ❤️

File: spec/system/backend/solidus_stripe/orders/payments_spec.rb [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+  include SolidusStripe::BackendTestHelper
+  stub_authorization!
+
+  context 'with successful payment operations' do
+    it 'navigates to the payments page' do
+      payment = creates_authorized_payment
+      visits_payments_page
+
+      payment_displayed(payment)
+    end
```
Comment (elia) [inline spec/system/backend/solidus_stripe/orders/payments_spec.rb]: Love this specs, really easy to follow them 🙌 

I just have a nit to point out, i.e. the methods are using a slightly different way of stating things, deviating from the usual RSpec/Capybara style. Some examples:
- `visits_payment_page` would usually be an imperative command, like `visit_payments_page`
- `payment_displayed` would usually express an expectation, like `expects(page).to display_payment` or `expects_page_to_display_payment`

File: spec/system/backend/solidus_stripe/orders/payments_spec.rb [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+    it 'displays log entries for a payment' do
+      payment = creates_authorized_payment
+      message = "authorized payment intent"
+      creates_payment_log_entry(payment, message)
```
Comment (elia) [inline spec/system/backend/solidus_stripe/orders/payments_spec.rb]: Thoughts on using factories for `creates_authorized_payment` and `creates_payment_log_entry`?
e.g. `create(:stripe_authorized_payment)`

**Review by elia — COMMENTED** [review]
Great to see this also finally covered by specs 🙌 

Left a few comments, mostly nits but I'm interested in your take

File: spec/system/backend/solidus_stripe/orders/payments_spec.rb (hunk 3 above)
Comment (rainerdema) [reply spec/system/backend/solidus_stripe/orders/payments_spec.rb]: I agree. Thinking about that, I would also make some changes to the specs related to the solidus_starter_frontend in a separate PR.

**Review by elia — APPROVED** [review]

File: spec/system/backend/solidus_stripe/orders/payments_spec.rb (hunk 3 above)
Comment (rainerdema) [reply spec/system/backend/solidus_stripe/orders/payments_spec.rb]: I created an issue to do the same even for the specs related to Solidus Starter Frontend: https://github.com/solidusio/solidus_stripe/issues/248

File: spec/system/backend/solidus_stripe/orders/payments_spec.rb (hunk 4 above)
Comment (rainerdema) [reply spec/system/backend/solidus_stripe/orders/payments_spec.rb]: ✔️  I tried to keep the factories generic for possible other future use, let me know what you think

File: spec/system/backend/solidus_stripe/orders/payments_spec.rb [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+      expects_page_to_display_payment(payment)
+    end
+
+    it 'displays log entries for a payment' do
+      payment = create(:stripe_payment, :captured, order: order, payment_method: payment_method)
+
+      visit_payments_page
+
+      visit_payment_page(payment)
```
Comment (elia) [inline spec/system/backend/solidus_stripe/orders/payments_spec.rb]: Another nit: if we `visit` pages without any expectation or check we could just go to the payment page. Alternatively it could be:

```rb
visit "/admin/orders/#{order.number}/payments"
click_on payment.number
```

Ensuring we navigate the site with "clicks" after the initial visit.

Wither way is fine by me 🙌

**Review by elia — APPROVED** [review]
Awesome, one final nit, not blocking ✅
